src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/interior_data_manager_impl.cc

```diff
@@ -0,0 +1,123 @@
+    : rc_plugin_(rc_plugin)
+    , cache_(cache)
+    , app_mngr_(app_mngr)
+    , rpc_service_(rpc_service) {}
+
+void InteriorDataManagerImpl::OnPolicyEvent(plugins::PolicyEvent event) {
+  UpdateHMISubscriptionsOnPolicyUpdated();
+}
+
+void InteriorDataManagerImpl::OnApplicationEvent(
+    plugins::ApplicationEvent event,
+    app_mngr::ApplicationSharedPtr application) {
+  if (plugins::ApplicationEvent::kApplicationUnregistered == event ||
+      plugins::ApplicationEvent::kApplicationExit == event) {
+    UpdateHMISubscriptionsOnAppUnregistered(*application);
+  }
+}
+
+void InteriorDataManagerImpl::OnDisablingRC() {
+  LOG4CXX_AUTO_TRACE(logger_);
+  auto existing_subscription = AppsSubscribedModules();
+  std::set<std::string> subscribed_modules;
+  for (auto& pair : existing_subscription) {
+    auto& app = pair.first;
+    auto rc_extension = RCHelpers::GetRCExtension(*app);
+    for (const auto& module : pair.second) {
+      subscribed_modules.insert(module);
+      rc_extension->UnsubscribeFromInteriorVehicleData(module);
+    }
+  }
+  for (auto& module : subscribed_modules) {
+    LOG4CXX_TRACE(logger_, "unsubscribe " << module);
+    UnsubscribeFromInteriorVehicleData(module);
+  }
+}
+
+void InteriorDataManagerImpl::UpdateHMISubscriptionsOnPolicyUpdated() {
+  auto apps_allowed_modules =
+      RCHelpers::GetApplicaitonsAllowedModules(app_mngr_);
+  auto apps_subscribed_modules = AppsSubscribedModules();
+  InteriorDataManagerImpl::AppsModules apps_disallowed_modules;
+  for (auto& pair : apps_subscribed_modules) {
+    auto& allowed = apps_allowed_modules[pair.first];
+    auto& subscribed = pair.second;
+    std::vector<std::string> disallowed_modules;
+    std::set_difference(subscribed.begin(),
+                        subscribed.end(),
+                        allowed.begin(),
+                        allowed.end(),
+                        std::back_inserter(disallowed_modules));
+    apps_disallowed_modules[pair.first] = disallowed_modules;
+  }
+
+  for (auto& pair : apps_disallowed_modules) {
+    auto& app = pair.first;
+    auto rc_extension = RCHelpers::GetRCExtension(*app);
+    for (const auto& module : pair.second) {
+      rc_extension->UnsubscribeFromInteriorVehicleData(module);
+      auto apps_subscribed =
+          RCHelpers::AppsSubscribedToModuleType(module, app_mngr_);
+      if (apps_subscribed.empty()) {
+        UnsubscribeFromInteriorVehicleData(module);
+      }
+    }
+  }
+}
+
+void InteriorDataManagerImpl::UpdateHMISubscriptionsOnAppUnregistered(
+    application_manager::Application& app) {
+  LOG4CXX_AUTO_TRACE(logger_);
+  auto rc_extension = RCHelpers::GetRCExtension(app);
+  auto subscribed_data = rc_extension->InteriorVehicleDataSubscriptions();
+  rc_extension->UnsubscribeFromInteriorVehicleData();
+  for (auto& data : subscribed_data) {
+    auto apps_subscribed = RCHelpers::AppsSubscribedTo(app_mngr_, data);
+    if (apps_subscribed.empty()) {
```

Comment:
@alexkutsan  do some investigation and either fix redundant check or provide explanation of the behavior

Reply:
@yang1070 this function will be called both for app exit and app unregistered. In case if app unregistered, there would be no apps_subscribed in the collection. but if the application was exited, the application will in the  apps_subscribed. So we should Check both conditions.